elrond-wasm/src/storage/mappers/linked_list_mapper.rs

```diff
@@ -0,0 +1,353 @@
+		storage_get(
+			self.api.clone(),
+			self.build_index_named_key(NODE_IDENTIFIER, index)
+				.as_slice(),
+		)
+	}
+
+	fn set_node(&self, index: u32, item: Node) {
+		storage_set(
+			self.api.clone(),
+			self.build_index_named_key(NODE_IDENTIFIER, index)
+				.as_slice(),
+			&item,
+		);
+	}
+
+	fn clear_node(&self, index: u32) {
+		storage_set(
+			self.api.clone(),
+			self.build_index_named_key(NODE_IDENTIFIER, index)
+				.as_slice(),
+			&BoxedBytes::empty(),
+		);
+	}
+
+	fn get_value(&self, index: u32) -> T {
+		storage_get(
+			self.api.clone(),
+			self.build_index_named_key(VALUE_IDENTIFIER, index)
+				.as_slice(),
+		)
+	}
+
+	fn get_value_option(&self, index: u32) -> Option<T> {
+		if index == NULL_ENTRY {
+			return None;
+		}
+		Some(self.get_value(index))
+	}
+
+	fn set_value(&self, index: u32, value: &T) {
+		storage_set(
+			self.api.clone(),
+			self.build_index_named_key(VALUE_IDENTIFIER, index)
+				.as_slice(),
+			value,
+		)
+	}
+
+	fn clear_value(&self, index: u32) {
+		storage_set(
+			self.api.clone(),
+			self.build_index_named_key(VALUE_IDENTIFIER, index)
+				.as_slice(),
+			&BoxedBytes::empty(),
+		)
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.get_len() == 0
+	}
+
+	pub fn len(&self) -> usize {
+		self.get_len() as usize
+	}
+
+	pub(crate) fn push_back_indexed(&mut self, elt: &T) -> u32 {
+		let new_index = self.generate_new_index();
+		let len = self.get_len();
+		let mut previous = NULL_ENTRY;
+		if len == 0 {
+			self.set_front(new_index);
+		} else {
+			let back = self.get_back();
+			let mut back_node = self.get_node(back);
+			back_node.next = new_index;
+			previous = back;
+			self.set_node(back, back_node);
+		}
+		self.set_node(
+			new_index,
+			Node {
+				previous,
+				next: NULL_ENTRY,
+			},
+		);
+		self.set_back(new_index);
+		self.set_value(new_index, &elt);
+		self.set_len(len + 1);
+		new_index
+	}
+
+	pub fn push_back(&mut self, elt: T) {
+		let _ = self.push_back_indexed(&elt);
+	}
+
+	pub fn push_front(&mut self, elt: T) {
+		let new_index = self.generate_new_index();
+		let len = self.get_len();
+		let mut next = NULL_ENTRY;
+		if len == 0 {
+			self.set_back(new_index);
+		} else {
+			let front = self.get_front();
+			let mut front_node = self.get_node(front);
+			front_node.previous = new_index;
+			next = front;
+			self.set_node(front, front_node);
+		}
+		self.set_node(
+			new_index,
+			Node {
+				previous: NULL_ENTRY,
+				next,
+			},
+		);
+		self.set_front(new_index);
+		self.set_value(new_index, &elt);
+		self.set_len(len + 1);
+	}
+
+	pub fn front(&self) -> Option<T> {
+		self.get_value_option(self.get_front())
+	}
+
+	pub fn back(&self) -> Option<T> {
+		self.get_value_option(self.get_back())
+	}
+
+	pub fn pop_back(&mut self) -> Option<T> {
+		self.remove_at_index(self.get_back())
+	}
+
+	pub fn pop_front(&mut self) -> Option<T> {
+		self.remove_at_index(self.get_front())
+	}
+
+	pub(crate) fn remove_at_index(&self, index: u32) -> Option<T> {
```

Comment:
Might also be worth mentioning in the comment that this is O(1). The rust linked list has the default O(n) remove.

Reply:
I renamed all usages of `index` to `node_id`, I agree it's more suggestive.
Added rust doc for public functions, with the special O(1) mention for the remove case.
Regarding the `remove_by_node_id` - I would leave it as `pub(crate)` because it would require the user to manage node ids. I would prefer to restrict the API in order to keep the code idiomatic. Managing IDs is usually a sign that another data structure is needed - which should be implemented in this crate.
Note that I added this method because it's going to be used by the follow-up pull request for the SetMapper, where the set is going to keep track of the `node_id`s of each element inserted into the collection. However, it adds the restriction that no duplicate elements are allowed, so they should stay as separate classes.
Reference, for how the remove is used in SetMapper: https://github.com/ElrondNetwork/elrond-wasm-rs/blob/set-and-map-mappers/elrond-wasm/src/storage/mappers/set_mapper.rs#L106